### ISSUE TITLE
Resolving dependency coordinates from another project in current Gradle build instead of from repository

### ISSUE THREAD

**Issue by rgoldberg** [gradle/gradle#5382]
Please note: I do not know the Gradle code base, so I might be using terms loosely, and/or making incorrect guesses about how Gradle works.

### Current Behavior
It seems that dependencies on libraries supplied by a repository are specified using coordinates, while dependencies on libraries supplied by other projects in the same build are specified using `project(':xyz')`.

### Expected Behavior
Is there currently any way (either built into Gradle or via a plugin) to resolve a dependency specified using coordinates to a project in the same build that provides an artifact matching those coordinates, instead of resolving to an artifact obtained from a repository?

Dependencies should always be specified in the same manner, regardless of whether the dependency is fulfilled by an artifact supplied by a repository, or by another project in the current build.

Why should a project know whether one of its dependencies is fulfilled by another project or by a repository?  Having two different styles seems to break standard rules of encapsulation, which causes problems described below in the Context section.

I'd imagine that such functionality would be provided by an implementation of `ArtifactRepository` and a corresponding method in `DefaultRepositoryHandler`, the latter possibly being named `build`, `gradleBuild`, `project`, or `gradleProject`. This "repository" could be put at the top of the `repositories` closure.

### Context
This would allow:

1. only cloning source for projects that must be modified, simplifying the file system, the IDE, and the build process.
2. easily redirecting a dependency on a third-party library (that is normally downloaded from a repo like maven central) to a forked version that has been included as a project in the build, without modifying the referring project at all.  I imagine that the forked project would have to be included in `settings.gradle`, though there should be some way to include every subdirectory that contains a `build.gradle` (either direct children only, or all recursive descendants) of a specified directory. (Do any such multi-directory includes exist?)
3. using a single consistent style of dependency specification, providing proper encapsulation

Building a dependency & deploying it to, e.g., a local maven repository doesn't solve my needs because I cannot edit the code for a maven artifact in my IDE while debugging and have my runtime reload from source. Also, I couldn't apply existing  breakpoints in source code open in my IDE to the equivalent classes from a maven artifact.  I'd also have two copies of the class in my IDE caches: one from my open project source, the other from the local maven repo.  Selecting the proper one each time would be burdensome and error-prone.

**Comment by Vampire** [comment]
What you are after is the composite build feature.
If projects are in one multi-projects build, they are tightly coupled and referenced by the `project(...)` syntax.
With the composite build feature (either via commandline properties, or also via `settings.gradle`), you can replace a binary dependency (`implementation 'foo:bar:1.0.0'`) by a "sub-build" that produces those artifacts with the given coordinates.
See https://docs.gradle.org/current/userguide/composite_builds.html for more information.

**Comment by rgoldberg** [comment]
Thanks for the info. 2 follow-up questions:

A) what is the best way to share settings (like versions of third-party libraries) between independent builds in a composite build? Should I have a Mercurial repo that is a subrepo of each of my Mercurial repos? Can I somehow include one gradle file from another within the same gradle project? e.g., say the subrepo is gradle-common, and contains libraries.gradle. Can I include gradle-common/libraries.gradle from my build.gradle in my project root?

B) is there anything like includeSiblingBuilds() for settings.gradle? This would include all gradle builds that are siblings of the root directory of the composite build. This would allow me to, e.g., clone the guava repo to a sibling directory, and have the composite build pick it up without me having to change includeBuild statements in the composite build. If this doesn’t exist, would a headless plugin be the best way to add such a method? If so, and if there are multiple types of plugins, what type of plugin must I create to provide this?

Thanks.

**Comment by Vampire** [comment]
> what is the best way to share settings (like versions of third-party libraries) between independent builds in a composite build?

Independent builds are just like you said independent.
If you want to have different version of libs in your fork, I guess the best way is to change it in the fork.
Maybe you could also create some gradle file or plugin that sets the versions you want,
or you can use dependency constraints if you want higher versions than are in the build. https://docs.gradle.org/current/userguide/userguide_single.html#sec:dependency_constraints
The question is why you need this.
If you need higher versions and also need code changes in the fork, just also change the version in the fork.
If it is just to use the newer version, dependency resolution will take the higher version in your including build anyway when constructing the dependency tree for the including build.

> Should I have a Mercurial repo that is a subrepo of each of my Mercurial repos?

How you manage repositories is totally up to you, that does not really influence the Gradle build, except if you expect the repo at a certain place and it is not there.

> Can I somehow include one gradle file from another within the same gradle project? e.g., say the subrepo is gradle-common, and contains libraries.gradle. Can build.gradle in my project root include gradle-common/libraries.gradle?

Yes you can, those are called script plugins. Just do `apply from: 'gradle-common/libraries.gradle'` or maybe `apply from: project(':gradle-common').file('libraries.gradle')`.

> is there anything like includeSiblingBuilds() for settings.gradle? This would include all gradle builds that are siblings of the root directory of the composite build. This would allow me to, e.g., clone the guava repo to a sibling directory, and have the composite build pick it up without me having to change includeBuild statements in my existing composite build. If this doesn’t exist, would a gradle plugin be the best way to add such functionality? If so, and if there are multiple types of plugins, what type of plugin must I create to provide this?

`settings.gradle` is a Gradle file like any other, so it is a Groovy script actually and you can do any Groovy magic you like, like listing sibling directories, searching for a `build.gradle` file and then include that build or similar.

**Comment by rgoldberg** [comment]
Thanks again for the additional info. To clarify:

A) I want to share library versions between my own projects, not with third-party forks. If a new version is released of a library that I’m using in multiple projects, I’d like to update the version in only one place, rather than having to change versions in multiple separate projects. If I use a mercurial subrepo like the proposed `gradle-common`, then I can update the version there. I can then run a script to pull & update `gradle-common` in every repo, and try building & testing. If all succeeds, I can then commit the subrepo updates. If anything fails, I can fix it, or easily revert every subrepo to the old version.

B) it sounds like there currently is no simple equivalent to my proposed `includeSiblingBuilds()`. I’m sure I can figure out how to implement such a method, but how should I package it to easily use it in multiple builds, and to provide it to the community at large? Would I make a gradle plugin, or must I use some other packaging scheme? If the former, and if there are multiple types of gradle plugins, which type should I use? Once I know the high-level packaging methodology, I’ll read the docs for it; I just don’t want to read all about gradle plugins only to find out that I should have been investigating something else.

**Comment by Vampire** [comment]
> I want to share library versions between my own projects, not with forks of third-party projects. If a new version is released of a library that I use in multiple projects, I’d like to update the version in only one place, rather than having to change versions in multiple separate projects. If I use a mercurial subrepo like the proposed gradle-common, then I can update the version there. I can then run a script to pull & update gradle-common in every repo, and try building & testing. If all succeeds, I can then commit the subrepo updates. If anything fails, I can fix it, or easily revert every subrepo to the old version. I plan to use this version-sharing scheme unless someone suggests something better; I don’t want to invent an inferior wheel if better ones already exist.

Ah, I see.
Then yes, what you plan sounds reasonable.
Alternatively you could make a BOM with your library versions and deploy it to some repository if you have one, but I guess this not the case for you.
So the sub-repo solution might be good, and there you could either add the deps if it is all the same for your projects, or set the dependency constraints as mentioned to just suggest the versions and then use version-less dependency declarations in your real projects.

> it sounds like there currently is no simple equivalent to my proposed includeSiblingBuilds() (if one does indeed exist, please correct me).

At least not that I know of.

> I’m sure I can figure out how to implement such a method, but how should I package it to easily use it in multiple builds,

You can do the same as you plan to do with the dependencies.

> and to provide it to the community at large? Would I make a gradle plugin, or must I use some other packaging scheme? If the former, and if there are multiple types of gradle plugins, which type should I use? Once I know the high-level packaging methodology, I’ll read the docs for it; I just don’t want to read all about gradle plugins only to find out that I should have been investigating something else.

Then you could either try to make a PR to Gradle to add such a method or write a `Plugin<Settings>` that does it and release it to the Plugin Portal, so yes, a plugin could do it.

Basically it would be just something like

```groovy
settingsDir.parentFile.eachDir {
    if (it.list { dir, name -> name == 'settings.gradle' }) {
        includeBuild it
    }
}
```

**Comment by rgoldberg** [comment]
Thanks again for all the info.  Yet another question:

It seems that I cannot just run `./gradlew build` on a composite build and have it build all included builds.  It seems that I need to add the following to the composite build's `build.gradle`:

```
task build {
	dependsOn gradle.includedBuild('includedProject1').task(':build')
	dependsOn gradle.includedBuild('includedProject2').task(':build')
	// ...
}

```

Is this the intended behavior, or am I doing something wrong, or is this a bug, or…?

If I run a task on a composite build that isn't explicitly defined in that composite build, is there any way to make Gradle attempt to run that task on each of the included builds?

If there's no built-in or plugin way to do this, do you think it would be possible for someone (possibly me)  to implement this?  I'd imagine that I'd need to register some sort of handler for undefined tasks, but I don't know if that's possible.

If not, I guess I that some plugin might be able to iterate over all tasks of all included projects, and dynamically define a task in the composite project for every found task name that isn't defined in the composite build, and have each such generated task call the same task on every included build that has said task.

**Comment by Vampire** [comment]
If you need taht, then you don't actually use the included builds.
With that declaration you just make the `build` task depend on the included `build` tasks, but that does not mean you use the artifacts built by the included projects.
Maybe the included build does not declare the same group / version than with which it is published, then it cannot match them automatically of course.
The included builds are run automatically if necessary.
Maybe you need additional dependency substitutions to map the dependencies.
When it usually is necessary is described in the Docs I linked you to.

**Comment by rgoldberg** [comment]
FYI: I'm using the Gradle Wrapper 4.7 with the Oracle Java SDK 10.0.1+10 on macOS 10.13.4.

If I don’t add that task definition to my composite build, then running the following in my composite build’s root finishes without reporting any errors, but doesn’t ever build anything (i.e. doesn't run javac on any of my included builds, doesn't create a build directory containing class files in any of my included build directories, etc.):

```
./gradlew build
```

Am I doing something wrong?

My composite build must be properly resolving coordinate-specified dependencies to included builds because, when I include the aforementioned build task definition and run `./gradlew build` with two included projects (my-app & my-util, where my-app has a dependency on the coordinates of my-util), both included builds build fine without any complaints. If I remove my-util from the composite build, then the composite build fails because it cannot satisfy my-app's dependency on my-util.

I derived my build task definition from the example in the composite build docs labelled "Example: Depending on task from included build".  The example was for the run task, but the docs didn't show any usage of the build task in a composite build.

**Comment by Vampire** [comment]
Why do you think "must be properly resolving coordinate-specified dependencies to included builds"?
I don't see your build so I can only guess, but if you e. g. have in `my-app` a dependency on `'foo:my-util:1.0.0'` and then include the build of `my-util` and `my-util' has not declared group `foo` and name `my-util` but something else, then it will not automatically replace your binary dependency, but manually making the `build` task of the composing build depending on the build task of the included build will still work fine. The problem is, that then the `build` task of the composing build will depend on the `build` task of the included build, but nothing more. `my-app` will not compile against the fork and will not use the fork. To achieve this if coordinates do not match, you need to configure the substitution manually as noted earlier.

**Comment by Vampire** [comment]
PS: if you use Java 10 it might be a good idea to upgrade to 4.8-rc-1, at least if you start your build with it it sates that it supports Java 10, so maybe something does not work properly with 4.7, I don't know though.

**Comment by Vampire** [comment]
Hm, that sounds strange then.
Are you sure they are not built?
Maybe you just don't see it because of the reduced output.
Try running your build with the `--console=verbose` argument to see what tasks actually are run.
Or try running with `--info` or `--debug` to find out what is strange.
You will never automatically have likewise-named tasks connected automatically as that would not make sense for composite builds, but if the JAR is needed, the `jar` task of the included build will be run to build it.

**Comment by Vampire** [comment]
Sure it does. 
If it needs the artifact of that build, it builds the artifact of that build. 
Otherwise your code that compiles against code on that artifact wouldn't be able to build.
If you want such a `build -> build` dependency, you have to create it yourself, but you can do it also automatically generically like you can include all builds you find automatically. 

It just doesn't make sense for the normal use-case, as in the normal use-case you don't want that dependency, but you want the artifact to be built when you need it and that is what is done.

Don't get me wrong, if you want such a dependency for your builds, go on and add it. I'm just saying that the Gradle guys will probably never have this built in, as it is not the use-case for that feature as far as I can tell.
